add unit tests for proxy

david415 opened the pull request:
Here we replace the explicit calls to Read and Write with io.Copy; we utilize a wait group to determine when both goroutines are done. Crash-only design; no shutdown code path so far.

david415 committed:
change source to go fmt format

david415 committed:
Replace netCopy with minimal closure with io.Copy; remove superfluous test

david415 committed:
Support SOCKS proxy to UNIX domain sockets

here we also get rid of global variables and move flag parsing to Main

david415 committed:
Encapsulate proxy logic inside TLSProxy struct type and method receivers

david415 committed:
Use an interface for SNI parsing

david415 committed:
Create struct type and interface for onion resolver logic

david415 committed:
Move all proxy logic to proxy.go

starius commented:
`bufferSize` is unused now.

david415 committed:
Add unfinished unit test

this includes mock struct types for the onion resolver and sni parser interfaces

starius commented on entry_proxy/main.go:
```diff
@@ -83,12 +66,17 @@ func processRequest(clientConn net.Conn) {
-	<-finished
+	var wg sync.WaitGroup
+	wg.Add(2)
+
```
IMHO, this empty line is not needed.

starius commented:
Tested locally with instructions from #3

starius commented on entry_proxy/proxy.go:
```diff
@@ -0,0 +1,108 @@
+			log.Printf("Unable to accept request: %s", err)
+		}
+	}
+
```
IMHO, this empty line is not needed.

starius commented:
IMHO, name `onion_resolver.go` is not ideal, because it does not resolve onions. Onion resolver would take onion and return something else (e.g., IP address). This module takes ordinary domain name and returns an onion.

david415 committed:
Add proxy dialer interface and finish writing a basic unit test for the proxy

david415 committed:
Rename to resolve_to_onion.go

starius commented:
I am not happy with two options (`-proxyAddr` and `-proxyNet`) rather than one (`-proxy`). I have two questions about this: 
- Tor does not work on unix domain socket anyway, does it?
- Is there a way to put unix domain socket as host into `url.URL`?

david415 commented:
i added a basic unit test for the proxy. it's kind of like an end to end test except that the sni parser, hostname to onion resolver and proxy dialer have been replaced with test/mock replacements for testing purposes.

starius commented on entry_proxy/proxy_test.go:
```diff
@@ -0,0 +1,31 @@
+	proxyNet := "tcp"
+	proxyAddr := "127.0.0.1:34627"
+
+	proxy := NewTLSProxy(443, proxyNet, proxyAddr)
```
It fails if 34627 port is not free.
See how `proxy_test.go` [is written](https://go.googlesource.com/net/+/master/proxy/proxy_test.go#53):

``` go
endSystem, err := net.Listen("tcp", "127.0.0.1:0")
// ...
if c, err := proxy.Dial("tcp", endSystem.Addr().String()); err != nil {
```

starius commented:
Should we test it with real SNI parser, mock resolver and real socks5 server like [socks5Gateway from proxy_test.go](https://go.googlesource.com/net/+/master/proxy/proxy_test.go#82)?

david415 commented:
yeah i agree the proxy url was nicer than having two options.

david415 commented:
looks like tor now supports unix domain sockets for SOCKS
https://trac.torproject.org/projects/tor/ticket/12585

but let's try to see if the proxy url format can have a unix socket file...

starius commented:
By the way, can we add "tor" to option(s) name(s) to make its purpose more clear?

starius commented:
> looks like tor now supports unix domain sockets for SOCKS

cool!

david415 commented:
i think it's better to have the unit test not exercise code in the sni parser or socks clients. those modules have their own tests already... although the test doesn't provide coverage for the hostname to onion resolver; it's probably ok if that component has no tests.

starius commented on entry_proxy/main.go:
```diff
@@ -10,99 +10,56 @@ See also:
+	"%{level:.4s} %{id:03x} %{message}",
+)
+var ttyFormat = logging.MustStringFormatter(
+	"%{color}%{time:15:04:05} ▶ %{level:.4s} %{id:03x}%{color:reset} %{message}",
```
IMHO, `s/▶ //` to remove non-ascii characters from a log.

starius commented:
What is the purpose of third-party logger?

Sample output of standard logger:

```
2016/08/07 14:31:47 test-ssl.pasta.cf was resolved to pokodugigmf66twh.onion
```

Sample output of third-party logger:

```
14:56:16 ▶ INFO 003 test-ssl.pasta.cf was resolved to pokodugigmf66twh.onion
```

IMHO, new format is not ideal:
- non-ascii character
- date is missing

david415 commented:
yeah that logger is totally optional. we could remove it.

the logger only logs non ascii to terminal otherwise it logs ascii to the log. it's pretty... various log levels such as debug and warn get printed on the terminal with different colors.

starius commented:
Currently all messages except `Unable to listen on` use `Info` level and are printed with default color. If we use this colourful logger, it makes sense to convert some messages to "warnings", e.g., "Unable to get target server name from SNI".

Can this logger write messages to different files? For instance, `nginx` writes "<=info" messages to `access.log` and ">=warn" messages to `error.log`.

david415 commented:
no it doesn't write to different files i don't think. but that's systemd's job. we could run this directly from the terminal or perhaps make a systemd unit file. probably if we want to get it packaged in debian it requires a unit file. we could log to our own files but it seems easier to let systemd/journalctl handle the logging.

starius commented:
> it seems easier to let systemd/journalctl handle the logging

agreed

david415 commented:
yeah i think i want to remove the fancy logger

starius commented on entry_proxy/proxy_test.go:
```diff
@@ -0,0 +1,213 @@
+
+	<-fakeTorListener.Received
+
+	want := "meow\n"
```
Can you create this variable before `fmt.Fprint(conn, "meow\r\n")` and reuse its value in it to prevent copy-paste?

Why these strings are not equal (`meow\r\n` and `meow\n`)?

david415 committed:
Remove fancy logger and use the standard golang library logger instead

starius commented:
By the way this tool should not write per-request logs in production (=default) mode. Otherwise somebody who has logs of TCP sessions of a user can take these logs and know which sites were visited.

I think, we should add a flag which explicitly enables such logs.

david415 committed:
Make test use variable for string comparison

david415 commented:
yes definitely.

starius commented:
Can you refactor `ResolveToOnion`, please?
I think, we should factor out function `txts -> onion` from it and cover this function with tests.

david415 commented:
oh i see what you mean; you are suggesting i have the test run using the real hostname to onion resolver but inside the resolver implementation we use an interface to handle the dns TXT queries?

starius commented:
> you are suggesting i have the test run using the real hostname to onion resolver but inside the resolver implementation we use an interface to handle the dns TXT queries

Not exactly this. I suggest something like this:

``` go
func parseTXT(txt []string) (onion string, err error) {
    if len(txts) == 0 {
        err = errors.New(fmt.Sprintf("No TXT records for %s", hostname))
        return
    }
    for _, txt := range txts {
        match := o.regex.FindStringSubmatch(txt)
        if match != nil {
            return match[2], nil // the submatch we are interested in
        }
    }
    return "", errors.New(fmt.Sprintf("No suitable TXT records for %s", hostname))
}

func (o *RealOnionResolver) ResolveToOnion(hostname string) (onion string, err error) {
    txts, err := net.LookupTXT(hostname)
    if err != nil {
        return
    }
    return parseTXT(txt)
}
```

and then test `parseTXT`.

david415 committed:
Add a TxtResolver unit test so that we can use instead of the old resolver interface

this allows us to gain great code coverage by the already written unit test

david415 commented:
hmm interesting that the unit tests work fine for me locally... but travis fails to run them.
https://travis-ci.org/DonnchaC/oniongateway/jobs/150412127

```
0.34s$ go test -v github.com/DonnchaC/oniongateway/entry_proxy -covermode=count -coverprofile=coverage.out

=== RUN TestTLSProxy-2

2016/08/07 12:50:19 starting listener service tcp:127.0.0.1:34628

--- FAIL: TestTLSProxy-2 (0.00 seconds)

    proxy_test.go:200: failed to connect to proxy: dial tcp 127.0.0.1:44333: connection refused

panic: runtime error: invalid memory address or nil pointer dereference [recovered]

    panic: runtime error: invalid memory address or nil pointer dereference

[signal 0xb code=0x1 addr=0x20 pc=0x47fa9d]

```

i think the proxy unit test has no need for real network io. we could probably find a way to test it using a mock connection that passes the io through to the other mock connection. or may

starius commented:
Tests fail locally here with the same error message.

starius committed:
fix tests: pick free port instead of hardcoded

starius commented:
I fixed test failures in https://github.com/david415/oniongateway/pull/1

I have not reviewed new tests carefully, but they seems overcomplicated now.

starius commented:
Some Travis jobs failed. What is wrong with `"meow\n"`?

david415 committed:
Replace Fprintf with Write in proxy unit test

david415 committed:
Fix previous commit variable reference

david415 committed:
Attempt to fix proxy unit test for older versions of golang

starius commented:
Could you print strings in hex format to discover bytes, please?

david415 commented:
that was a weird failure. fixed now.

starius commented:
So the fix was in the commit https://github.com/DonnchaC/oniongateway/pull/4/commits/1039265b347eff921d09f94cfab43207824e94f2
Can you explain this issue in details, please?

david415 committed:
Make proxy unit test print hex strings for debugging messages

david415 commented:
It seems TrimSpace is not consistent across all versions of golang.

starius commented:
IMHO comparing trimmed text is overkill, because the proxy guarantees "memcmp"-level equality of its input and output. It worth adding stress test passing garbage data through the proxy to make sure it works as expected.

david415 commented:
yeah the trim was removed. i was using some of that code for a different project involving a line protocol.

david415 committed:
Add a couple of minor code cleanup corrections from golint

david415 committed:
Remove unused import of "errors"

david415 commented:
hmm i suppose this is ready to merge... but the git history is rather sloppy. we could also squash some of the commits to make it few number of commits.

starius commented:
I am increasing coverage of `ResolveToOnion`.

starius committed:
entry_proxy: more tests for resolve_to_onion_test

Cover uncovered `if`s in
https://coveralls.io/builds/7341558/source?filename=entry_proxy%2Fresolve_to_onion.go

starius commented:
Done: https://github.com/david415/oniongateway/pull/2

starius commented:
> we could also squash some of the commits to make it few number of commits.

+1 for things reverted/fixed up/redone in this very pull request.

david415 commented:
new pull request here with identical code changes but a few of my commits are squashed or reworded
https://github.com/DonnchaC/oniongateway/pull/11

starius commented:
Tested with Unix domain socket.

To setup Tor in use Unix domain socket, set `SOCKSPort unix:/path/to/file`
To run `entry_proxy` with Unix domain socket, invoke it with `-proxyAddr=/path/to/file -proxyNet=unix`
